cpp/src/arrow/compute/exec/plan_test.cc

```diff
@@ -322,26 +306,16 @@ void TestSourceSink(
     std::string source_factory_name,
     std::function<Result<std::vector<ElementType>>(const BatchesWithSchema&)>
         to_elements) {
-  ASSERT_OK_AND_ASSIGN(auto executor, arrow::internal::ThreadPool::Make(1));
-  ExecContext exec_context(default_memory_pool(), executor.get());
-  ASSERT_OK_AND_ASSIGN(auto plan, ExecPlan::Make(exec_context));
-  AsyncGenerator<std::optional<ExecBatch>> sink_gen;
-
   auto exp_batches = MakeBasicBatches();
   ASSERT_OK_AND_ASSIGN(auto elements, to_elements(exp_batches));
   auto element_it_maker = [&elements]() {
     return MakeVectorIterator<ElementType>(elements);
   };
-
-  ASSERT_OK(Declaration::Sequence({
-                                      {source_factory_name,
-                                       OptionsType{exp_batches.schema, element_it_maker}},
-                                      {"sink", SinkNodeOptions{&sink_gen}},
-                                  })
-                .AddToPlan(plan.get()));
-
-  ASSERT_THAT(StartAndCollect(plan.get(), sink_gen),
-              Finishes(ResultWith(UnorderedElementsAreArray(exp_batches.batches))));
+  Declaration plan(source_factory_name,
+                   OptionsType{exp_batches.schema, element_it_maker});
+  ASSERT_OK_AND_ASSIGN(auto result,
+                       DeclarationToExecBatches(std::move(plan), /*use_threads=*/false));
```

Comment:
Any particular reason for `use_threads=false` here?

Reply:
Not really, just left the test case as it hasn’t used parallel mode. Should we prefer testing both modes at all times? 

PS: I sort of did that, but again removed that thinking I am introducing which is not in the current test cases.

Reply:
I am setting this to default just now... I will confirm one's it is tested.

Reply:
Just for the curiosity I repeated this on the `use_threads=false` setting (which is the same state this was run before making this change, I guess).

The following was my observation

```bash
/home/asus/github/fork/arrow/cpp/src/arrow/compute/exec/plan_test.cc:887: Failure
Value of: sink_gen()
Expected: value value value is equal to ExecBatch
    # Rows: 4
    0: Array[-2147483380,808065407,-681488731,829937304]
    1: Array[true,false,false,false]

  Actual: 16-byte object <F0-5B 46-2E D1-55 00-00 90-75 47-2E D1-55 00-00>, whose value (ExecBatch
    # Rows: 4
    0: Array[-2147483418,113924762,1252144620,-193012790]
    1: Array[false,true,true,true]
) doesn't match, whose value ExecBatch
    # Rows: 4
    0: Array[-2147483418,113924762,1252144620,-193012790]
    1: Array[false,true,true,true]
 doesn't match
Google Test trace:
/home/asus/github/fork/arrow/cpp/src/arrow/compute/exec/plan_test.cc:866: single threaded
/home/asus/github/fork/arrow/cpp/src/arrow/compute/exec/plan_test.cc:863: unslowed


-----------------------------------------------------------------------------------------------------

/home/asus/github/fork/arrow/cpp/src/arrow/compute/exec/plan_test.cc:887: Failure
Value of: sink_gen()
Expected: value value value is equal to ExecBatch
    # Rows: 4
    0: Array[-2147483380,808065407,-681488731,829937304]
    1: Array[true,false,false,false]

  Actual: 16-byte object <00-66 F7-B1 AD-55 00-00 40-DE F7-B1 AD-55 00-00>, whose value (ExecBatch
    # Rows: 4
    0: Array[-2147483522,-417100609,768701078,-1848250513]
    1: Array[true,false,true,true]
) doesn't match, whose value ExecBatch
    # Rows: 4
    0: Array[-2147483522,-417100609,768701078,-1848250513]
    1: Array[true,false,true,true]
 doesn't match
Google Test trace:
/home/asus/github/fork/arrow/cpp/src/arrow/compute/exec/plan_test.cc:866: parallel
/home/asus/github/fork/arrow/cpp/src/arrow/compute/exec/plan_test.cc:863: unslowed


-----------------------------------------------------------------------------------------------------

/home/asus/github/fork/arrow/cpp/src/arrow/compute/exec/plan_test.cc:887: Failure
Value of: sink_gen()
Expected: value value value is equal to ExecBatch
    # Rows: 4
    0: Array[-2147483380,808065407,-681488731,829937304]
    1: Array[true,false,false,false]

  Actual: 16-byte object <B0-75 1A-B3 48-56 00-00 30-06 1D-B3 48-56 00-00>, whose value (ExecBatch
    # Rows: 4
    0: Array[-2147483522,-417100609,768701078,-1848250513]
    1: Array[true,false,true,true]
) doesn't match, whose value ExecBatch
    # Rows: 4
    0: Array[-2147483522,-417100609,768701078,-1848250513]
    1: Array[true,false,true,true]
 doesn't match
Google Test trace:
/home/asus/github/fork/arrow/cpp/src/arrow/compute/exec/plan_test.cc:866: parallel
/home/asus/github/fork/arrow/cpp/src/arrow/compute/exec/plan_test.cc:863: unslowed
```

For 160 repetitions (bash on 8 runs), I observed 3 failures. Could say 3/8 fails. Do you have a clue why this could be happening? The failing test is `ExecPlanExecution.StressSourceSinkStopped`. A single run includes all test cases. 

But when run just this case over 100 repetitions, it failed only 2 times. I suspect a resource allocation issue. Appreciate your thoughts.

Reply:
Ah, I have a fix for that test here: https://github.com/apache/arrow/pull/15253/files#diff-b4e59a9cedddaaa33c03f220c03ac812f9ebdfb06692c9f91c5f8f5988363b57R853 

However, I don't know if that PR will get finished in time for the release.  I'll open an independent PR with the fix.  The test does this:

 * Generate a bunch of random batches
 * Start the plan
 * Wait for one batch to arrive
 * Cancel/abort the plan
 * Verify the one batch that arrived was the first of the random batches

However, there is no guarantee, when run in parallel, that the first batch to complete will be the first batch generated.

Since the test doesn't fully consume the plan we can't use DeclarationToXyz here.  This means when run serially it is using "single-thread mode" which is still slightly parallel enough to cause random ordering.

In my fix I changed it so parallel=false would run synchronously (and have dependent order) and parallel=true would not require it to be the first batch generated.

Reply:
Shall we hold this until this PR is merged?

Reply:
No, sorry.  I thought you were modifying the test you were referring to but I see now you are not.  I think this PR is good to go.